boards: Pine64 PineTime DevKit0

sdorre opened the pull request:
Board definition for PineTime DevKit0 by Pine64 vendor. It is based on information published in https://wiki.pine64.org/index.php/PineTime

Based on initial work of Rafa Couto caligari@treboada.net

[PR still open](https://github.com/zephyrproject-rtos/zephyr/pull/19973)

rafacouto committed:
boards: PineTime board skeleton

Pinout reference: http://files.pine64.org/doc/datasheet/pinetime/PineTime%20Port%20Assignment%20rev0.2.pdf
Adding PineTime SPI and I2C, some DTS and configuration definitions.
Adding led and button gpios.

Signed-off-by: Rafa Couto <caligari@treboada.net>

rafacouto committed:
boards: pinetime_devkit0 basic sample.

Basic sample program that uses the one led and button from the Pinetime.
Led will turn on everytime the button is pushed.

Signed-off-by: Rafa Couto <caligari@treboada.net>

SuperPrower committed:
board: Improve pinetime_devkit0 configuration

Disable setting CONFIG_SOC_FAMILY_NRF manually
Limit SPI frequency to 8MHz
configure new st7789v in device tree

Signed-off-by: Sergey Koziakov <dya.eshshmai@gmail.com>

sdorre commented:
I think, I have know a good starting point. The sample app is running. I could use most of the GPIOs.
It may improve as I (and hopefully other users) start develop against that platform.

ioannisg reviewed (changes requested):
Looks ok, but I think the documentation should be enhanced, to include at least some info about the SoC and Board devices, building, flashing, debugging, etc.

ioannisg commented:
@SuperPrower  @rafacouto mind take a looks, since this PR contains commits you're signing-off?

ioannisg reviewed (approved):
LGTM

ioannisg commented:
@anangl wonder if you could take a quick look

carlescufi commented on samples/boards/pine64_pinetime/src/main.c:
```diff
@@ -0,0 +1,89 @@
+	while (1) {
+		/* button is pressed ==> turn on status LED */
+		u32_t val = 0u;
+		u8_t newVal = 0u;
```
Please get rid of CamelCase

anangl commented on boards/arm/pinetime_devkit0/pinetime_devkit0_defconfig:
```diff
@@ -0,0 +1,33 @@
+CONFIG_NRFX_TWIM0=y
+
+# enable SPI1 driver
+CONFIG_NRFX_SPI=y
+CONFIG_NRFX_SPI1=y
```
I guess it was supposed to be `CONFIG_SPI=y` instead. Or is the nrfx SPI driver indeed intended to be used directly (with [this](https://github.com/zephyrproject-rtos/hal_nordic/blob/master/nrfx/drivers/include/nrfx_spi.h) API), not through the spi_nrfx_spi shim and [Zephyr SPI API](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/spi.h)?

sdorre replied:
fixed it.

rafacouto commented on boards/arm/pinetime_devkit0/pinetime_devkit0.dts:
```diff
@@ -0,0 +1,201 @@
+	mosi-pin = <3>;
+	miso-pin = <4>;
+
+	cs-gpios = <&gpio0 5 0>, <&gpio0 25 0>;
```
You are right: CS pins are active-low.

sdorre replied:
fixed all GPIOS

rafacouto commented on boards/arm/pinetime_devkit0/pinetime_devkit0.dts:
```diff
@@ -0,0 +1,201 @@
+		compatible = "hynitron,cst816s";
+		reg = <0x15>;
+		label = "CST816S";
+		irq-gpios = <&gpio0 28 0>;
+		rst-gpios = <&gpio0 10 0>;
```
> Aren't these two pins active low?

Checked with [working firmware](https://github.com/JF002/Pinetime) and yes, they are active low.

anangl replied:
Then the `GPIO_ACTIVE_LOW` flag should be used instead of `0` as the third item in both those arrays.
And similarly for other active low pins referred in this file.

sdorre commented on boards/arm/pinetime_devkit0/pinetime_devkit0_defconfig:
```diff
@@ -0,0 +1,33 @@
+CONFIG_I2C=y
+CONFIG_I2C_0=y
+CONFIG_NRFX_TWIM=y
+CONFIG_NRFX_TWIM0=y
```
Good point, I fixed it.

anangl commented on samples/boards/pine64_pinetime/src/main.c:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	/* Pinetime trick : Enable button */
+	gpio_pin_set(dev_led, DT_GPIO_PIN(DT_ALIAS(sw1), gpios), 1u);
```
`dev_led` here is a bit confusing, as the button state is set, and the button pin was configured using `dev_btn`.
But since there is only one GPIO peripheral in nRF52832, `dev_led` and `dev_btn` will always point to the same device (perhaps then it would be better to use `dev_gpio` instead of those two).
You can also use `GPIO_OUTPUT_ACTIVE` in the above call to `gpio_pin_configure()` and avoid the need of calling `gpio_pin_set()`.

anangl reviewed (commented):
Perhaps it would be good to squash together commits that add/modify the same thing.

carlescufi commented:
@sdorre could you please squash the commits into a meaningful set? Thank you!

sdorre commented:
> @sdorre could you please squash the commits into a meaningful set? Thank you!

Can I squash my commit with the initial commit which were not mine ? 
I just don't want to remove any credit for the initial draft.

Which copyright should I put in the header ? I am not completely the author of this work.

carlescufi commented:
> > @sdorre could you please squash the commits into a meaningful set? Thank you!
> 
> @carlescufi Can I squash my commit with the initial commit which were not mine ?
> I just don't want to remove any credit for the initial draft.

No, in that case split them by author as well. Just a reasonable set that doesn't include amends to commits.

> @anangl Which copyright should I put in the header ? I am not completely the author of this work.

Add only yours if none other is present.

sdorre committed:
samples: Adapt Pinetime sample program

Adopt latest Zephyr syntax
Improve project configuration
Improve sample code

Signed-off-by: Stephane Dorre <stephane.dorre@gmail.com>

sdorre committed:
boards: Adapt Pinetime board configuration

Move i2c to defconfig
Add documentation
Improve dts
Remove unnecessary option and only use the default ones.
Provide a default Active state directly in the device tree
Fix copyright

Signed-off-by: Stephane Dorre <stephane.dorre@gmail.com>

sdorre commented:
@carlescufi  Is that better now ?

sdorre commented:
@nashif @galak @MaureenHelm 
Is there something missing to merge that PR now ?

carlescufi commented:
> @nashif @galak @MaureenHelm
> Is there something missing to merge that PR now ?

No, this looks good now. I will merge it.

martinjaeger commented:
I'm not sure if something went wrong with the squashing of the commits in the end, but the code of this sample does not compile and also the board documentation has some flaws. Found this coincidentally when I searched for the legacy int types in the code base.

CI probably didn't detect the compile issues because of missing sample.yaml.

Submitted PR #30083 as a fix.
